java-checks/src/main/java/org/sonar/java/checks/LazyArgEvaluationCheck.java

```diff
@@ -0,0 +1,183 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.CatchTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Rule(key = "S2629")
+public class LazyArgEvaluationCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+
+  static final class Methods {
+
+    static final class Guava {
+      static final MethodMatcher PRECONDITIONS = MethodMatcher.create()
+        .name("checkState").withAnyParameters();
+
+      private Guava() {}
+    }
+
+    static final class Logging {
+      static final MethodMatcher LOG = MethodMatcher.create()
+        .name("log").addParameter(TypeCriteria.anyType()).addParameter(JAVA_LANG_STRING);
+      static final MethodMatcher TRACE = slf4jPrototype().name("trace");
+      static final MethodMatcher DEBUG = slf4jPrototype().name("debug");
+      static final MethodMatcher INFO = slf4jPrototype().name("info");
+      static final MethodMatcher WARN = slf4jPrototype().name("warn");
+      static final MethodMatcher ERROR = slf4jPrototype().name("error");
+
+      private Logging() {}
+
+      static MethodMatcher slf4jPrototype() {
+        return MethodMatcher.create().addParameter(JAVA_LANG_STRING).addParameter(TypeCriteria.anyType());
+      }
+    }
+
+    private Methods() {}
+  }
+
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    if (context.getSemanticModel() == null) {
+      return;
+    }
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    for (MethodMatcher methodMatcher : getMethodInvocationMatchers()) {
+      if (methodMatcher.matches(tree)) {
+        onMethodInvocationFound(tree);
+        break;
+      }
+    }
+  }
+
+  @Override
+  public void visitCatch(CatchTree tree) {
+    // cut the visit on catch blocks, because we don't mind some performance loss on exceptional paths
```

Comment:
Okay so that's the reason why you don't want to rely on `AbstractMethodDetection`. But then what about this case (nested code in catch blocks):
```
MyClass myField;

public void exceptionalPaths() {
  try {
    myField = tryGetMyField();
  } catch (Exception e) {
    LOG.info("Something went wrong: " + message); // ignored, - OK -
    myField = new MyClass() {
      @Overidde
      void doSomethingAllTheTime() {
        LOG.info("logging all the time consuming resources for nothing " + computeValue() + generateStuff()); // consequently ignored, but which should raise an issue
      }
    };
  }
}
```